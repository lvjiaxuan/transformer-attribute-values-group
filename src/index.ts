import type MagicString from 'magic-string'
import MagicStringStack from 'magic-string-stack'
import type { SourceCodeTransformer } from 'unocss'

export function main(_code: MagicString) {
  const code = new MagicStringStack(_code.toString())

  // no combinator
  code.replace(
    /(?<=class="[\s\S]*?)&\[([\s\S]+?)=\(([^)(]+?)\)(?=[\s\S]*?")/g,
    (_match, variant: string, values: string) => values
      .split(/\s+/g)
      .filter(Boolean)
      .map((value, idx, arr) => `&[${variant}=${value}${arr.length - 1 === idx ? '' : ']'}`)
      .join(','),
  )

  // with combinator
  code.replace(
    /(?<=class="[\s\S]*?)\[(\S+?)&\[([\s\S]+?)=\(([\s\S]+?)\)\](\S*?)\](?=[\s\S]*?")/g,
    (_match, combinatorAhead: string, variant: string, values: string, combinatorBehind: string) => `${values
      .split(/\s+/g)
      .filter(Boolean)
      .map((value, idx, arr) => {
        return `${idx === 0 ? '[' : ''}${combinatorAhead}&[${variant}=${value}${combinatorBehind ? `]${combinatorBehind}` : (arr.length - 1 === idx ? '' : ']')}`
      })
      .join(',')}]${combinatorBehind ? '' : ']'}`,
  )

  // data-attribute
  const matches = code.toString().matchAll(/(?<=class="[\s\S]*?)(\S+:)*?data-\[\S+=\((?<values>[^\)]+)\)\]:([^(]\S+|\([^\)]+\))(?=[\s\S]*?")/g)
  ;[...matches].forEach((match) => {
    const values = match?.groups?.values as unknown as string

    const joinIt = values
      ?.split(/\s/g)
      .filter(Boolean)
      .reduce((acc, value, _idx) => {
        acc.push(match[0].replaceAll(`(${values})`, value))
        return acc
      }, [] as string[])

    code.replace(match[0], joinIt.join(' '))
    code.commit()
  })

  return code.toString()
}

export default {
  name: 'unocss-transformer-attribute-values-group',
  enforce: 'pre',
  transform: (code) => {
    main(code)
  },
} satisfies SourceCodeTransformer
