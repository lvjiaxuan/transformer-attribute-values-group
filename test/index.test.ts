import { describe, expect, it } from 'vitest'
import MagicString from 'magic-string'
import { main } from '../src'

const m = (s: string) => new MagicString(`<div class="${s}" />`)

describe('group attribute values', () => {
  it('basic', () => {
    expect(
      main(m('[&[type=(number text)]]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[&[type=number],&[type=text]]:c-red" />"`)
  })

  it('with operator', () => {
    expect(
      main(m('[&[type^=(number text)]]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[&[type^=number],&[type^=text]]:c-red" />"`)
  })

  it('with combinator', () => {
    expect(
      // Next-sibling combinator
      main(m('[.foo+&[type=(number text)]]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[.foo+&[type=number],.foo+&[type=text]]:c-red" />"`)

    expect(
      // Child combinator
      main(m('[.foo>&[type=(number text)]]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[.foo>&[type=number],.foo>&[type=text]]:c-red" />"`)

    expect(
      // Column combinator
      main(m('[.foo||&[type=(number text)]]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[.foo||&[type=number],.foo||&[type=text]]:c-red" />"`)

    expect(
      // Subsequent sibling combinator
      main(m('[.foo~&[type=(number text)]]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[.foo~&[type=number],.foo~&[type=text]]:c-red" />"`)

    expect(
      // Descendant combinator
      main(m('[.foo_&[type=(number text)]]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[.foo_&[type=number],.foo_&[type=text]]:c-red" />"`)

    expect(
      // Namespace separator
      main(m('[.foo|&[type=(number text)]]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[.foo|&[type=number],.foo|&[type=text]]:c-red" />"`)

    expect(
      // a sequence of combinator
      main(m('[.foo+&[type=(number text)]~.bar_.sim]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[.foo+&[type=number]~.bar_.sim,.foo+&[type=text]~.bar_.sim]:c-red" />"`)
  })

  it('multiple variant sorting', () => {
    expect(
      main(m('dark:hover:[&[type=(number text)]]:c-red')),
      // Note: unocss can parse this result. but it doesn't seem good.
      // Just ignore it because it's too special.
      // .dark
      //   .dark\:hover\:\[\&\[type\=number\]\,\&\[type\=text\]\]\:c-red[type='number'],
      // .dark\:hover\:\[\&\[type\=number\]\,\&\[type\=text\]\]\:c-red[type='text']:hover {
      //   --un-text-opacity: 1;
      //   color: rgb(248 113 113 / var(--un-text-opacity));
      // }
    ).toMatchInlineSnapshot(`"<div class="dark:hover:[&[type=number],&[type=text]]:c-red" />"`)
  })

  it('compound class selectors', () => {
    expect(
      main(m('[.foo&[type=(number text)]]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[.foo&[type=number],.foo&[type=text]]:c-red" />"`)

    expect(
      main(m('[.foo&[type=(number text)].bar]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[.foo&[type=number].bar,.foo&[type=text].bar]:c-red" />"`)
  })

  it('multiple', () => {
    expect(
      main(m('[&[type=(number text)]]:c-red [&[size=(large small middle)]]:p-1')),
    ).toMatchInlineSnapshot(`"<div class="[&[type=number],&[type=text]]:c-red [&[size=large],&[size=small],&[size=middle]]:p-1" />"`)
  })

  it('multiple in the same arbitrary-variants', () => {
    expect(
      main(m('[&[type=(number text)],&[aa=(bb cc)]]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[&[type=number],&[type=text)],&[aa=(bb],&[type=cc]]:c-red" />"`)
  })

  it('with variant group', () => {
    expect(
      main(m('[&[type=(number text)]]:(c-red m-1)')),
    ).toMatchInlineSnapshot(`"<div class="[&[type=number],&[type=text]]:(c-red m-1)" />"`)

    expect(
      main(m('[&[type=(number text)],&[aa=(bb cc)]]:(c-red m-1)')),
    ).toMatchInlineSnapshot(`"<div class="[&[type=number],&[type=text)],&[aa=(bb],&[type=cc]]:(c-red m-1)" />"`)
  })

  it('with newline', () => {
    expect(
      main(m('[&[type=(number\n text)],&[aa=(bb cc)]]:(\n  c-red \n     m-1\n     p-1\n  )')),
    ).toMatchInlineSnapshot(`"<div class="[&[type=number],&[type=text)],&[aa=(bb],&[type=cc)]]:(],&[type=c-red],&[type=m-1],&[type=p-1" />"`)
  })

  it('empty group', () => {
    expect(
      main(m('[&[type=()]]:p-1')),
    ).toMatchInlineSnapshot(`"<div class="[&[type=()]]:p-1" />"`)
  })
})

describe.skip('group data-attribute values', () => {
  it('basic', () => {
    expect(
      main(m('data-[xxx=(foo bar)]:p-1')),
    ).toMatchInlineSnapshot(`"<div class="data-[xxx=foo]:p-1" /> <div class="data-[xxx=bar]:p-1" />"`)
  })

  it('with operator', () => {
    expect(
      main(m('data-[xxx^=(foo bar)]:p-1')),
    ).toMatchInlineSnapshot(`"<div class="data-[xxx^=foo]:p-1" /> <div class="data-[xxx^=bar]:p-1" />"`)
  })

  it.todo('multiple', () => {
    expect(
      main(m('data-[xxx=(foo bar)]:p-1 data-[yyy=(aaa bbb)]:m-1')),
    ).toMatchInlineSnapshot()
  })

  it('multiple variant sorting', () => {
    expect(
      main(m('dark:hover:data-[xxx=(foo bar)]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="dark:hover:data-[xxx=foo]:c-red" /> <div class="dark:hover:data-[xxx=bar]:c-red" />"`)
  })

  it('with variant group', () => {
    expect(
      main(m('data-[xxx=(foo bar)]:(p-1 m-1)')),
    ).toMatchInlineSnapshot(`"<div class="data-[xxx=foo]:(p-1 m-1)" /> <div class="data-[xxx=bar]:(p-1 m-1)" />"`)
  })

  it('with newline', () => {
    expect(
      main(m('data-[xxx=(foo\n bar)]:(\n  c-red \n     m-1\n     p-1\n  )')),
    ).toMatchInlineSnapshot(`
      "<div class="data-[xxx=foo]:(
        c-red 
           m-1
           p-1
        )" /> <div class="data-[xxx=bar]:(
        c-red 
           m-1
           p-1
        )" />"
    `)
  })

  it('empty', () => {
    expect(
      main(m('data-[xxx=()]:p-1')),
    ).toMatchInlineSnapshot(`"<div class="data-[xxx=()]:p-1" />"`)
  })
})

describe.skip('random combination', () => {
  it.todo('combination 1', () => expect(
    main(m('data-[xxx=(foo bar)]:(p-1 m-1) data-[yyy=(aa)]:p-1')),
  ).toMatchInlineSnapshot())

  it.todo('combination 2', () => {
    expect(main(m('[&[type=(number text)]]:c-red data-[xxx=(foo bar)]:p-1')))
      .toMatchInlineSnapshot()

    expect(main(m('data-[xxx=(foo bar)]:p-1 [&[type=(number text)]]:c-red')))
      .toMatchInlineSnapshot()
  })

  it('with grouping selector and combinator', () => {
    expect(
      main(m('[.foo,&[type=(number text)]~.bar]:c-red')),
    ).toMatchInlineSnapshot(`"<div class="[.foo,&[type=number],&[type=text]~.bar]:c-red" />"`)
  })
})
