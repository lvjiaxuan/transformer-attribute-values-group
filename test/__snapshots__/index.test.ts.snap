// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`references some special cases from UnoCSS > basic > "![&[type=(number text)]]:(m-2 p-2)" 1`] = `
"<div class="p1 ![&[type=number],&[type=text]]:(m-2 p-2) m1" />
<p class="p2 ![&[type=number],&[type=text]]:(m-2 p-2) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "!data-[xxx=(foo bar)]:(m-2 p-2)" 1`] = `
"<div class="p1 !data-[xxx=foo]:(m-2 p-2) data-[xxx=bar]:(m-2 p-2) m1" />
<p class="p2 !data-[xxx=foo]:(m-2 p-2) data-[xxx=bar]:(m-2 p-2) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "<sm:hover:[&[type=(number text)]]:(p-1 p-2)" 1`] = `
"<div class="p1 <sm:hover:[&[type=number],&[type=text]]:(p-1 p-2) m1" />
<p class="p2 <sm:hover:[&[type=number],&[type=text]]:(p-1 p-2) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "<sm:hover:data-[xxx=(foo bar)]:(p-1 p-2)" 1`] = `
"<div class="p1 <sm:hover:data-[xxx=foo]:(p-1 p-2) <sm:hover:data-[xxx=bar]:(p-1 p-2) m1" />
<p class="p2 <sm:hover:data-[xxx=foo]:(p-1 p-2) <sm:hover:data-[xxx=bar]:(p-1 p-2) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "[&[type=(number text)]]:(
!m-2 
p-2
)" 1`] = `
"<div class="p1 [&[type=number],&[type=text]]:(
!m-2 
p-2
) m1" />
<p class="p2 [&[type=number],&[type=text]]:(
!m-2 
p-2
) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "[&[type=(number text)]]:(!m-2 p-2)" 1`] = `
"<div class="p1 [&[type=number],&[type=text]]:(!m-2 p-2) m1" />
<p class="p2 [&[type=number],&[type=text]]:(!m-2 p-2) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "[&[type=(number text)]]:(grid grid-cols-[1fr,50%])" 1`] = `
"<div class="p1 [&[type=number],&[type=text]]:(grid grid-cols-[1fr,50%]) m1" />
<p class="p2 [&[type=number],&[type=text]]:(grid grid-cols-[1fr,50%]) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "[&[type=(number text)]]:(w-1/2 h-[calc(100%-4rem)])" 1`] = `
"<div class="p1 [&[type=number],&[type=text]]:(w-1/2 h-[calc(100%-4rem)]) m1" />
<p class="p2 [&[type=number],&[type=text]]:(w-1/2 h-[calc(100%-4rem)]) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "[&[type=(number text)]]:(w-40vw pr-4.5rem)" 1`] = `
"<div class="p1 [&[type=number],&[type=text]]:(w-40vw pr-4.5rem) m1" />
<p class="p2 [&[type=number],&[type=text]]:(w-40vw pr-4.5rem) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "data-[xxx=(foo bar)]:(
!m-2 
p-2
)" 1`] = `
"<div class="p1 data-[xxx=foo]:(
!m-2 
p-2
) data-[xxx=bar]:(
!m-2 
p-2
) m1" />
<p class="p2 data-[xxx=foo]:(
!m-2 
p-2
) data-[xxx=bar]:(
!m-2 
p-2
) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "data-[xxx=(foo bar)]:(!m-2 p-2)" 1`] = `
"<div class="p1 data-[xxx=foo]:(!m-2 p-2) data-[xxx=bar]:(!m-2 p-2) m1" />
<p class="p2 data-[xxx=foo]:(!m-2 p-2) data-[xxx=bar]:(!m-2 p-2) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "data-[xxx=(foo bar)]:(grid grid-cols-[1fr,50%])" 1`] = `
"<div class="p1 data-[xxx=foo]:(grid grid-cols-[1fr,50%]) data-[xxx=bar]:(grid grid-cols-[1fr,50%]) m1" />
<p class="p2 data-[xxx=foo]:(grid grid-cols-[1fr,50%]) data-[xxx=bar]:(grid grid-cols-[1fr,50%]) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "data-[xxx=(foo bar)]:(w-1/2 h-[calc(100%-4rem)])" 1`] = `
"<div class="p1 data-[xxx=foo]:(w-1/2 h-[calc(100%-4rem) data-[xxx=bar]:(w-1/2 h-[calc(100%-4rem)]) m1" />
<p class="p2 data-[xxx=foo]:(w-1/2 h-[calc(100%-4rem) data-[xxx=bar]:(w-1/2 h-[calc(100%-4rem)]) m2" />
"
`;

exports[`references some special cases from UnoCSS > basic > "data-[xxx=(foo bar)]:(w-40vw pr-4.5rem)" 1`] = `
"<div class="p1 data-[xxx=foo]:(w-40vw pr-4.5rem) data-[xxx=bar]:(w-40vw pr-4.5rem) m1" />
<p class="p2 data-[xxx=foo]:(w-40vw pr-4.5rem) data-[xxx=bar]:(w-40vw pr-4.5rem) m2" />
"
`;
